src/Protocol/ForgeHandshake.cpp

```diff
@@ -0,0 +1,304 @@
+
+// ForgeHandshake.cpp
+
+// Implements Forge protocol handshaking
+
+#include "Globals.h"
+#include "ForgeHandshake.h"
+#include "json/json.h"
+#include "../Server.h"
+#include "../Bindings/PluginManager.h"
+#include "../ClientHandle.h"
+#include "../Root.h"
+
+cForgeHandshake::cForgeHandshake(cClientHandle *client) : m_IsForgeClient(false), m_Errored(false), m_Client(client)
+{
+}
+
+
+
+
+
+void cForgeHandshake::AugmentServerListPing(Json::Value & a_ResponseValue)
+{
+	UInt32 ProtocolVersion = m_Client->GetProtocolVersion();
+	cForgeMods & Mods = cRoot::Get()->GetServer()->GetRegisteredForgeMods(ProtocolVersion);
+
+	if (!Mods.HasMods())
+	{
+		return;
+	}
+
+	LOG("Received server ping from version: %d", ProtocolVersion);
+
+	// modinfo:
+	Json::Value Modinfo;
+	Modinfo["type"] = "FML";
+
+	Json::Value ModList(Json::arrayValue);
+
+	for (auto & item: Mods.GetMods())
+	{
+		Json::Value Mod;
+		Mod["modid"] = item.first;
+		Mod["version"] = item.second;
+		ModList.append(Mod);
+	}
+
+	Modinfo["modList"] = ModList;
+
+	// Augment the response:
+	a_ResponseValue["modinfo"] = Modinfo;
+}
+
+
+
+
+
+void cForgeHandshake::BeginForgeHandshake(const AString & a_Name, const AString & a_UUID, const Json::Value & a_Properties)
+{
+	ASSERT(m_IsForgeClient);
+
+	m_Name = &a_Name;
+	m_UUID = &a_UUID;
+	m_Properties = &a_Properties;
+
+	AStringVector Channels = { "FML|HS", "FML", "FML|MP", "FML", "FORGE" };
+	AString ChannelsString;
+
+	for (AStringVector::iterator itr = Channels.begin(); itr != Channels.end(); ++itr)
+	{
+		ChannelsString.append(*itr);
+		ChannelsString.push_back('\0');
+	}
+
+	m_Client->SendPluginMessage("REGISTER", ChannelsString);
+	SendServerHello();
+}
+
+
+
+
+
+void cForgeHandshake::SendServerHello()
+{
+	AString Message;
+	// Discriminator | Byte | Always 0 for ServerHello
+	Message.push_back(Discriminator_ServerHello);
+	// FML protocol Version | Byte | Determined from NetworkRegistery. Currently 2.
+	Message.push_back('\2');
+	// Dimension TODO
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+	Message.push_back('\0');
+
+	m_Client->SendPluginMessage("FML|HS", Message);
+}
+
+AStringMap cForgeHandshake::ParseModList(const char * a_Data, size_t a_Size)
+{
+	cByteBuffer Buf(a_Size);
+	Buf.Write(a_Data, a_Size);
+
+	Int8 Discriminator;
+	Buf.ReadBEInt8(Discriminator);
+
+	ASSERT(Discriminator == 2);
```

Comment:
Do not assert. Check extensively, consider all client code broken-by-default. :)